extensions/aerosol_optics/mo_aerosol_optics.F90

```diff
@@ -0,0 +1,633 @@
+                                    tau_bcar_table, ssa_bcar_table, asy_bcar_table, &
+                                    tau_ocar_rh_table, ssa_ocar_rh_table, asy_ocar_rh_table, &
+                                    tau_ocar_table, ssa_ocar_table, asy_ocar_table, &
+                                    tau, taussa, taussag)
+
+    integer,                               intent(in) :: ncol, nlay, nbnd, nbin, nrh
+    logical(wl), dimension(ncol,nlay),     intent(in) :: mask
+    integer,     dimension(ncol,nlay),     intent(in) :: type
+    real(wp),    dimension(ncol,nlay),     intent(in) :: size, mass, rh
+
+    real(wp),    dimension(nbin),          intent(in) :: aero_bin_lo, aero_bin_hi
+    real(wp),    dimension(nrh),           intent(in) :: aero_rh
+    integer,     dimension(2),             intent(in) :: rh_nsteps
+    real(wp),    dimension(2),             intent(in) :: rh_step_sizes, rh_offsets
+
+    real(wp),    dimension(nbin,    nbnd), intent(in) :: tau_dust_table, ssa_dust_table, asy_dust_table
+    real(wp),    dimension(nbin,nrh,nbnd), intent(in) :: tau_salt_table, ssa_salt_table, asy_salt_table
+    real(wp),    dimension(     nrh,nbnd), intent(in) :: tau_sulf_table, ssa_sulf_table, asy_sulf_table
+    real(wp),    dimension(     nrh,nbnd), intent(in) :: tau_bcar_rh_table, ssa_bcar_rh_table, asy_bcar_rh_table
+    real(wp),    dimension(         nbnd), intent(in) :: tau_bcar_table, ssa_bcar_table, asy_bcar_table
+    real(wp),    dimension(     nrh,nbnd), intent(in) :: tau_ocar_rh_table, ssa_ocar_rh_table, asy_ocar_rh_table
+    real(wp),    dimension(         nbnd), intent(in) :: tau_ocar_table, ssa_ocar_table, asy_ocar_table
+
+    real(wp),    dimension(ncol,nlay,nbnd)            :: tau, taussa, taussag
+    ! ---------------------------
+    integer  :: icol, ilay, ibnd, ibin
+    integer  :: index
+    real(wp) :: fint
+    real(wp) :: t, ts, tsg  ! tau, tau*ssa, tau*ssa*g
+    ! ---------------------------
+    !$acc parallel loop gang vector default(present) collapse(3)
+    !$omp target teams distribute parallel do simd collapse(3)
+    do ibnd = 1, nbnd
+      do ilay = 1,nlay
+        do icol = 1, ncol
+          ibin = get_aerosol_size_bin(nbin, aero_bin_lo, aero_bin_hi, size(icol,ilay))
+          if (mask(icol,ilay) .and. rh(icol,ilay) .le. rh_offsets(2)) then
+            index = min(floor((rh(icol,ilay) - rh_offsets(1))/rh_step_sizes(1))+1, rh_nsteps(1)-1)
+            fint = (rh(icol,ilay) - rh_offsets(1))/rh_step_sizes(1) - (index-1)
+          elseif (mask(icol,ilay) .and. rh(icol,ilay) .gt. rh_offsets(2)) then
+            index = min(floor((rh(icol,ilay) - rh_offsets(2))/rh_step_sizes(2))+1, rh_nsteps(2)-1)
+            fint = (rh(icol,ilay) - rh_offsets(2))/rh_step_sizes(2) - (index-1)
+          endif
+          ! dust
+          if (mask(icol,ilay) .and. type(icol,ilay) == merra_aero_dust) then
+            t   = mass(icol,ilay) * tau_dust_table(ibin,ibnd)
```

Comment:
@mjiacono There's a lot of repeated code here. Why not make two interpolation routines, one that depends on RH and the other only on size, then pass table and index to each?

Reply:
And following up, why interpolation to tau, tau*ssa, tau*ssa*g, and then normalize in a separate step? Since only one aerosol type is allowed per cell, why not compute the desired output directly and write it into the output arrays right away?

Reply:
@RobertPincus Interpolation is only done on the rh dimension. For size, the appropriate bin is selected, since there are only five bins of unequal extent. Only four of the seven aerosol types are interpolated on rh:  salt is 3-D (size, rh, band), and sulfate, bcar_rh, and ocar_rh are 2-D (rh, band). I can add 2-D and 3-D functions to interpolate those tables, but it’s not clear if you prefer calling these functions where needed within the existing loops on band, layer, and column, or pre-interpolating (with the mask) before these loops, putting the result in new arrays, and using those interpolated table values to define the intermediate optics.  

Regarding your second comment, you’ll recall that the aerosol extension is agnostic to longwave or shortwave and so has to handle 1-stream (absorption) or 2-stream (extinction) optics, which it currently does in the separate step. Do you propose building that complexity into the interpolation step?

Reply:
@mjiacono Thanks for the explanations and the good questions. They prompted me to look a bit more carefully at the treatment of RH and make some suggestions. 

Re interpolation: It seems like you are currently computing both the size index and the rh index and weight for every cell that contains any aerosols. That seems like a good strategy. 

My suggestion was that you write an interpolation function like: 
```
function linear_interp(table, index, weight) result(val)
  real(wp), dimension(:), intent(in) :: table 
  integer, intent(in) :: index 
  real(wp), intent(in) :: weight
  real(wp) :: val 
  
  val = table(index) + weight * (table(index+1) - table(index)))
  end function linear_interp
```
And use this in your `compute_all_from_table` function, i.e. 
```
t   = mass(icol,ilay) * interp(tau_salt_table(ibin, :, ibnd))
```
The : notation extracts the slice of the `tau_salt_table` corresponding to the rh dimension at the values of `ibin` and `ibnd`. This would be more efficient if you reordered the tables to have the RH dimension first (then you'd be passing continuous bits of memory to the function). 

Is that clear?

Reply:
@mjiacono Re: the number of tables, you could make the first dimension of every table be `quantity` of length 3. Your interpolation routine could then return a vector of length 3 instead of a scalar. 

If you were being asked for absorption optical depth only it's true that you'd only make use of the first two elements. But the cost of interpolating the extra quantities is tiny compared to the gain in simplicity.

Reply:
@RobertPincus Thanks for clarifying your preferred direction, and yes, I'll reduce the number of tables.